pkg/consistenthash/maglev/maglev.go

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package maglev
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+
+	cluster_v2 "kmesh.net/kmesh/api/v2/cluster"
+	"kmesh.net/kmesh/api/v2/endpoint"
+	"kmesh.net/kmesh/pkg/utils/hash"
+)
+
+const (
+	DefaultTableSize    uint64 = 16381
+	DefaultHashSeed            = "JLfvgnHc2kaSUFaI"
+	MaglevOuterMapName         = "outer_of_maglev"
+	MaglevInnerMapName         = "inner_of_maglev"
+	MaglevMapMaxEntries        = 65536
+	ClusterNameMaxLen          = 192
+)
+
+var (
+	outer           *ebpf.Map
+	seedMurmur      uint32
+	maglevTableSize uint64
+)
+
+type Backend struct {
+	ep     *endpoint.Endpoint
+	index  int
+	offset uint64
+	skip   uint64
+	next   uint64
+}
+
+func InitMaglevMap() error {
+	maglevTableSize = DefaultTableSize
+	opt := &ebpf.LoadPinOptions{}
+
+	outer_map, err := ebpf.LoadPinnedMap("/sys/fs/bpf"+"/bpf_kmesh/map/"+MaglevOuterMapName, opt)
+	if err != nil {
+		return fmt.Errorf("load outer map of maglev failed err: %v", err)
+	}
+	outer = outer_map
+
+	d, err := base64.StdEncoding.DecodeString(DefaultHashSeed)
+	if err != nil {
+		return fmt.Errorf("cannot decode base64 Maglev hash seed %q: %w", DefaultHashSeed, err)
+	}
+	if len(d) != 12 {
+		return fmt.Errorf("decoded hash seed is %d bytes (not 12 bytes)", len(d))
+	}
+	seedMurmur = uint32(d[0])<<24 | uint32(d[1])<<16 | uint32(d[2])<<8 | uint32(d[3])
+
```

Comment:
What is the  code for?

Reply:
For the func getOffsetAndSkip to compute hash.